File: android/src/main/kotlin/com/palmstreet/braintree/flutter_braintree_plugin/FlutterBraintreePluginHelper.kt

package com.palmstreet.braintree.flutter_braintree_plugin

import com.braintreepayments.api.PayPalAccountNonce
import com.braintreepayments.api.PayPalCreditFinancing
import com.braintreepayments.api.PayPalCreditFinancingAmount
import com.braintreepayments.api.PayPalLineItem
import com.braintreepayments.api.PostalAddress
import com.braintreepayments.api.VenmoAccountNonce
import com.braintreepayments.api.VenmoLineItem
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result


class FlutterBraintreePluginHelper {
    companion object {
        fun getAuthorization(call: MethodCall): String? {
            val clientToken = string("clientToken", call)
            val tokenizationKey = string("tokenizationKey", call)
            val authorizationKey = string("authorization", call)

            return clientToken ?: tokenizationKey ?: authorizationKey
        }

        fun buildPayPalPaymentNonceDict(nonce: PayPalAccountNonce?): Map<String, Any?>? {
            nonce ?: return null
            return mapOf(
                "nonce" to nonce.string,
                "type" to "PayPal",
                "isDefault" to nonce.isDefault,
                "email" to nonce.email,
                "firstName" to nonce.firstName,
                "lastName" to nonce.lastName,
                "phone" to nonce.phone,
                "clientMetadataID" to nonce.clientMetadataId,
                "payerID" to nonce.payerId,
                "billingAddress" to buildPostalAddressDict(nonce.billingAddress),
                "shippingAddress" to buildPostalAddressDict(nonce.shippingAddress),
                "creditFinancing" to buildPayPalCreditFinancingDict(nonce.creditFinancing),
                "authenticateUrl" to nonce.authenticateUrl
            )
        }

        fun buildVenmoPaymentNonceDict(nonce: VenmoAccountNonce?): Map<String, Any?>? {
            nonce ?: return null
            return mapOf(
                "nonce" to nonce.string,
                "type" to "Venmo",
                "isDefault" to nonce.isDefault,
                "email" to nonce.email,
                "externalId" to nonce.externalId,
                "firstName" to nonce.firstName,
                "lastName" to nonce.lastName,
                "phoneNumber" to nonce.phoneNumber,
                "username" to nonce.username,
                "billingAddress" to buildPostalAddressDict(nonce.billingAddress),
                "shippingAddress" to buildPostalAddressDict(nonce.shippingAddress)
            )
        }

        private fun buildPostalAddressDict(address: PostalAddress?): Map<String, Any?>? {
            address ?: return null
            return mapOf(
                "recipientName" to address.recipientName,
                "streetAddress" to address.streetAddress,
                "extendedAddress" to address.extendedAddress,
                "locality" to address.locality,
                "countryCodeAlpha2" to address.countryCodeAlpha2,
                "postalCode" to address.postalCode,
                "region" to address.region
            )
        }

        private fun buildPayPalCreditFinancingDict(financing: PayPalCreditFinancing?): Map<String, Any?>? {
            financing ?: return null
            return mapOf(
                "cardAmountImmutable" to financing.isCardAmountImmutable,
                "monthlyPayment" to buildPayPalCreditFinancingAmountDict(financing.monthlyPayment),
                "payerAcceptance" to financing.hasPayerAcceptance(),
                "term" to financing.term,
                "totalCost" to buildPayPalCreditFinancingAmountDict(financing.totalCost),
                "totalInterest" to buildPayPalCreditFinancingAmountDict(financing.totalInterest)
            )
        }

        private fun buildPayPalCreditFinancingAmountDict(financing: PayPalCreditFinancingAmount?): Map<String, Any?>? {
            financing ?: return null
            return mapOf(
                "currency" to financing.currency,
                "value" to financing.value
            )
        }

        fun makeVenmoItems(from: List<Any>): List<VenmoLineItem>? {
            val venmoItems = from as? List<Map<String, Any>> ?: return null
            val outList = mutableListOf<VenmoLineItem>()
            for (venmoItem in venmoItems) {
                val quantity = venmoItem["quantity"] as? Int ?: return null
                val unitAmount = venmoItem["unitAmount"] as? String ?: return null
                val name = venmoItem["name"] as? String ?: return null
                val kind = venmoItem["kind"] as? Int ?: return null
                val item = VenmoLineItem(kind.toString(), name, quantity, unitAmount)
                val unitTaxAmount = venmoItem["unitTaxAmount"] as? String
                if (unitTaxAmount != null) {
                    item.setUnitTaxAmount(unitTaxAmount)
                }
                val itemDescription = venmoItem["itemDescription"] as? String
                if (itemDescription != null) {
                    item.setDescription(itemDescription)
                }
                val productCode = venmoItem["productCode"] as? String
                if (productCode != null) {
                    item.setProductCode(productCode)
                }
                val url = venmoItem["url"] as? String
                if (url != null) {
                    item.setUrl(url)
                }
                outList.add(item)
            }
            return outList
        }

        fun makePayPalItems(from: List<Any>): List<PayPalLineItem>? {
            val paypalItems = from as? List<Map<String, Any>> ?: return null
            val outList = mutableListOf<PayPalLineItem>()
            for (paypalItem in paypalItems) {
                val quantity = paypalItem["quantity"] as? Int ?: return null
                val unitAmount = paypalItem["unitAmount"] as? String ?: return null
                val name = paypalItem["name"] as? String ?: return null
                val kind = paypalItem["kind"] as? Int ?: return null
                val item = PayPalLineItem(kind.toString(), name, quantity.toString(), unitAmount)
                val unitTaxAmount = paypalItem["unitTaxAmount"] as? String
                if (unitTaxAmount != null) {
                    item.setUnitTaxAmount(unitTaxAmount)
                }
                val itemDescription = paypalItem["itemDescription"] as? String
                if (itemDescription != null) {
                    item.setDescription(itemDescription)
                }
                val productCode = paypalItem["productCode"] as? String
                if (productCode != null) {
                    item.setProductCode(productCode)
                }
                val url = paypalItem["url"] as? String
                if (url != null) {
                    item.setUrl(url)
                }
                outList.add(item)
            }
            return outList
        }

        fun returnAuthorizationMissingError(result: Result) {
            returnFlutterError(
                result,
                "braintree_error",
                "Authorization not specified (no clientToken or tokenizationKey)"
            )
        }

        fun returnFlutterError(result: Result, code: String, message: String) {
            result.error(code, message, null)
        }

        fun string(key: String, call: MethodCall): String? {
            return (call.arguments as? Map<String, Any>)?.get(key) as? String
        }

        fun bool(key: String, call: MethodCall): Boolean? {
            return (call.arguments as? Map<String, Any>)?.get(key) as? Boolean
        }

        fun dict(key: String, call: MethodCall): Map<String, Any>? {
            return (call.arguments as? Map<String, Any>)?.get(key) as? Map<String, Any>
        }
    }
}

File: ios/Classes/FlutterBraintreePluginHelper.swift

import Foundation
import Flutter
import Braintree

class FlutterBraintreePluginHelper {

  class func getAuthorization(call: FlutterMethodCall) -> String? {
    let clientToken = string(for: "clientToken", in: call)
    let tokenizationKey = string(for: "tokenizationKey", in: call)
    let authorizationKey = string(for: "authorization", in: call)

    guard let authorization = clientToken
            ?? tokenizationKey
            ?? authorizationKey else {
      return nil
    }

    return authorization
  }

  class func buildPaymentNonceDict(nonce: BTPaymentMethodNonce?) -> [String: Any?]? {
    guard let nonce = nonce else { return nil }
    var dict = [String: Any?]()
    dict["nonce"] = nonce.nonce
    dict["type"] = nonce.type
    dict["isDefault"] = nonce.isDefault

    if let paypalNonce = nonce as? BTPayPalAccountNonce {
      dict["email"] = paypalNonce.email
      dict["firstName"] = paypalNonce.firstName
      dict["lastName"] = paypalNonce.lastName
      dict["phone"] = paypalNonce.phone
      dict["clientMetadataID"] = paypalNonce.clientMetadataID
      dict["payerID"] = paypalNonce.payerID
      dict["billingAddress"] = buildPostalAddressDict(address: paypalNonce.billingAddress)
      dict["shippingAddress"] = buildPostalAddressDict(address: paypalNonce.shippingAddress)
      dict["creditFinancing"] = buildPayPalCreditFinancingDict(financing: paypalNonce.creditFinancing)
    }
    if let venmoNonce = nonce as? BTVenmoAccountNonce {
      dict["email"] = venmoNonce.email
      dict["externalId"] = venmoNonce.externalID
      dict["firstName"] = venmoNonce.firstName
      dict["lastName"] = venmoNonce.lastName
      dict["phoneNumber"] = venmoNonce.phoneNumber
      dict["username"] = venmoNonce.username
      dict["billingAddress"] = buildPostalAddressDict(address: venmoNonce.billingAddress)
      dict["shippingAddress"] = buildPostalAddressDict(address: venmoNonce.shippingAddress)
    }
    return dict
  }

  class func buildPostalAddressDict(address: BTPostalAddress?) -> [String: Any?]? {
    guard let address = address else { return nil }
    var dict = [String: Any?]()
    dict["recipientName"] = address.recipientName
    dict["streetAddress"] = address.streetAddress
    dict["extendedAddress"] = address.extendedAddress
    dict["locality"] = address.locality
    dict["countryCodeAlpha2"] = address.countryCodeAlpha2
    dict["postalCode"] = address.postalCode
    dict["region"] = address.region
    return dict
  }

  class func buildPayPalCreditFinancingDict(financing: BTPayPalCreditFinancing?) -> [String: Any?]? {
    guard let financing = financing else { return nil }
    var dict = [String: Any?]()
    dict["cardAmountImmutable"] = financing.cardAmountImmutable
    dict["monthlyPayment"] = buildPayPalCreditFinancingAmountDict(financing: financing.monthlyPayment)
    dict["payerAcceptance"] = financing.payerAcceptance
    dict["term"] = financing.term
    dict["totalCost"] = buildPayPalCreditFinancingAmountDict(financing: financing.totalCost)
    dict["totalInterest"] = buildPayPalCreditFinancingAmountDict(financing: financing.totalInterest)
    return dict
  }

  class func buildPayPalCreditFinancingAmountDict(financing: BTPayPalCreditFinancingAmount?) -> [String: Any?]? {
    guard let financing = financing else { return nil }
    var dict = [String: Any?]()
    dict["currency"] = financing.currency
    dict["value"] = financing.value
    return dict
  }

  class func makeVenmoItems(from: [Any]) -> [BTVenmoLineItem]? {
    guard let venmoItems = from as? [[String: Any]] else {
      return nil;
    }

    var outList: [BTVenmoLineItem] = []
    for venmoItem in venmoItems {
      guard let quantity = venmoItem["quantity"] as? Int else {
        return nil;
      }
      guard let unitAmount = venmoItem["unitAmount"] as? String else {
        return nil;
      }
      guard let name = venmoItem["name"] as? String else {
        return nil;
      }
      guard let kind = venmoItem["kind"] as? Int else {
        return nil;
      }
      guard let venmoKind = BTVenmoLineItemKind.init(rawValue: kind) else {
        return nil;
      }
      let item = BTVenmoLineItem(quantity: quantity, unitAmount: unitAmount, name: name, kind: venmoKind)

      item.unitTaxAmount = venmoItem["unitTaxAmount"] as? String
      item.itemDescription = venmoItem["itemDescription"] as? String
      item.productCode = venmoItem["productCode"] as? String
      if let urlStr = venmoItem["url"] as? String, let url = URL.init(string: urlStr) {
        item.url = url
      }

      outList.append(item);
    }

    return outList;
  }

  class func makePayPalItems(from: [Any]) -> [BTPayPalLineItem]? {
    guard let paypalItems = from as? [[String: Any]] else {
      return nil;
    }

    var outList: [BTPayPalLineItem] = []
    for paypalItem in paypalItems {
      guard let quantity = paypalItem["quantity"] as? Int else {
        return nil;
      }
      guard let unitAmount = paypalItem["unitAmount"] as? String else {
        return nil;
      }
      guard let name = paypalItem["name"] as? String else {
        return nil;
      }
      guard let kind = paypalItem["kind"] as? Int else {
        return nil;
      }
      guard let paypalKind = BTPayPalLineItemKind.init(rawValue: kind) else {
        return nil;
      }
      let item = BTPayPalLineItem(quantity: "\(quantity)", unitAmount: unitAmount, name: name, kind: paypalKind)
      outList.append(item);
    }

    return outList;
  }


  class func returnAuthorizationMissingError (result: FlutterResult) {
    returnFlutterError(result: result, code: "braintree_error", message: "Authorization not specified (no clientToken or tokenizationKey)")
  }

  class func returnBraintreeError(result: FlutterResult, error: Error) {
    returnFlutterError(result: result, code: "braintree_error", message: error.localizedDescription)
  }

  class func returnFlutterError(result: FlutterResult, code: String, message: String) {
    result(FlutterError(code: code, message: message, details: nil));
  }

  class func string(for key: String, in call: FlutterMethodCall) -> String? {
    return (call.arguments as? [String: Any])?[key] as? String
  }

  class func bool(for key: String, in call: FlutterMethodCall) -> Bool? {
    return (call.arguments as? [String: Any])?[key] as? Bool
  }

  class func dict(for key: String, in call: FlutterMethodCall) -> [String: Any]? {
    return (call.arguments as? [String: Any])?[key] as? [String: Any]
  }
}
